engines/syguspdr.cpp

```diff
@@ -0,0 +1,1542 @@
+      const ProofGoal * pg = proof_goals.top();
+
+      if (!pg->idx) {
+        // went all the way back to initial
+        // need to create a new proof goal that's not managed by the queue
+        SygusPdr::reconstruct_trace(pg, cex_);
+
+        // in case this is spurious, clear the queue of proof goals
+        // which might not have been precise
+        // TODO might have to change this if there's an algorithm
+        // that refines but can keep proof goals around
+        proof_goals.clear();
+
+        return false;
+      }
+
+      if (is_blocked(pg)) {
+        logger.log(3,
+                   "Skipping already blocked proof goal <{}, {}>",
+                   pg->target.term,
+                   pg->idx);
+        // remove the proof goal since it has already been blocked
+        assert(pg == proof_goals.top());
+        proof_goals.pop();
+        continue;
+      }
+
+      IC3Formula collateral;  // populated by rel_ind_check
+      if (rel_ind_check(pg->idx, pg->target, collateral, true)) {
+        // this proof goal can be blocked
+        assert(!solver_context_);
+        assert(collateral.term);
+        logger.log(
+            3, "Blocking term at frame {}: {}", pg->idx, pg->target.term);
+
+        // remove the proof goal now that it has been blocked
+        assert(pg == proof_goals.top());
+        proof_goals.pop();
+
+        if (options_.ic3_indgen_) {
+          collateral = inductive_generalization(pg->idx, collateral);
+        } else {
+          // just negate the term
+          collateral = ic3formula_negate(collateral);
+        }
+
+        size_t idx = find_highest_frame(pg->idx, collateral);
+        assert(idx >= pg->idx);
+
+        assert(collateral.disjunction);
+        assert(collateral.term);
+        assert(collateral.children.size());
+        SygusPdr::constrain_frame(idx, collateral);
+
+        // re-add the proof goal at a higher frame if not blocked
+        // up to the frontier
+        if (idx < frontier_idx()) {
+          assert(!pg->target.disjunction);
+          proof_goals.new_proof_goal(pg->target, idx + 1, pg->next);
+        }
+
+      } else {
+        // could not block this proof goal
+        assert(collateral.term);
+        if (collateral.term == solver_true_)
+          ; // do nothing // because we need to extend it to 0 to 
+        proof_goals.new_proof_goal(collateral, pg->idx - 1, pg);
+      }
+    }  // end while(!proof_goals.empty())
+
+    assert(!(goal = IC3Formula()).term);  // in debug mode, reset it
+  }                                       // end while(reaches_bad(goal))
+
+  assert(proof_goals.empty());
+  return true;
+} // block_all
+
+// -----------------------------------------------------------------
+// allow input in the constraint (when there are additional constraints)
+// -----------------------------------------------------------------
+
+
+static bool subsumes(const IC3Formula &a, const IC3Formula &b)
```

Comment:
If `constrain_frame` isn't overridden, this doesn't need to be either right? I'd like to move towards that -- it's better not to duplicate.